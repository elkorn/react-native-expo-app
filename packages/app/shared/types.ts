export enum LoadingState {
  Idle = "Idle",
  Loading = "Loading",
  Success = "Success",
  Failure = "Failure",
}
