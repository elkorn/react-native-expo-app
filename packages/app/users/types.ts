import * as UsersActions from "./usersActions";
import { ActionType } from "typesafe-actions";
import { LoadingState } from "../shared/types";

export type UsersAction = ActionType<typeof UsersActions>;

export interface User {
  id: string;
  name: string;
  avatar: string;
}

export interface UsersState {
  loadingState: LoadingState;
  users?: User[];
  error?: Error;
}
