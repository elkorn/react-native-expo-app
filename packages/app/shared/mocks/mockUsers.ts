export const mockUsers = [
  {
    id: "1",
    avatar: " https://i.pravatar.cc/150?img=1",
    name: "Morgan James",
  },
  {
    id: "2",
    avatar: "https://i.pravatar.cc/150?img=2",
    name: "Esme Coleman",
  },
  {
    id: "3",
    avatar: "https://i.pravatar.cc/150?img=3",
    name: "Gary Watson",
  },
];
