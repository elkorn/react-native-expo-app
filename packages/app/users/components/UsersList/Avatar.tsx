import React, { useEffect, useRef, useState } from "react";
import { ActivityIndicator, Animated, StyleSheet, View } from "react-native";
import { LoadingState } from "../../../shared/types";

export default function Avatar({ src }: AvatarProps) {
  const [loading, setLoading] = useState(LoadingState.Idle);
  const fadeValue = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    if (loading === LoadingState.Success) {
      Animated.spring(fadeValue, {
        toValue: 1,
        useNativeDriver: true,
      }).start();
    }
  }, [loading]);

  return (
    <View style={styles.container}>
      <Animated.Image
        style={{ ...styles.image, opacity: fadeValue }}
        onLoadStart={() => setLoading(LoadingState.Loading)}
        onLoad={() => setLoading(LoadingState.Success)}
        source={{
          uri: src.trim(),
        }}
      />
      <ActivityIndicator
        hidesWhenStopped
        animating={loading === LoadingState.Loading}
      />
    </View>
  );
}

interface AvatarProps {
  src: string;
}

const shape = {
  borderRadius: 100,
  width: 100,
  height: 100,
};

const styles = StyleSheet.create({
  container: {
    position: "relative",
    justifyContent: "center",
    alignItems: "center",
    marginRight: 12,
    ...shape,
  },
  image: {
    ...shape,
    position: "absolute",
    top: 0,
    left: 0,
    backgroundColor: "#eee",
  },
  error: {},
});
