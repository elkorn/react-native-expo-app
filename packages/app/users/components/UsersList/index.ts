export { UsersListContainer as default } from "./UsersListContainer";
