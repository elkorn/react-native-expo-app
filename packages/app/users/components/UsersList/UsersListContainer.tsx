import React, { useEffect } from "react";
import { useDispatch, useSelector } from "react-redux";
import { fetchUsers } from "../../usersThunks";
import { UsersList } from "./UsersList";
import { State } from "../../../shared/state";

export function UsersListContainer() {
  const dispatch = useDispatch();
  // no useCallback since we don't have fetch arguments
  // and want to allow refreshing without making assumptions
  // on how the backend works.
  const dispatchFetchUsers = dispatch.bind(null, fetchUsers());

  useEffect(() => {
    dispatch(fetchUsers());
  }, []);

  const users = useSelector((state: State) => state.users.users);
  const isFetching = useSelector((state: State) => state.users.isFetching);

  return (
    <UsersList
      users={users}
      isFetching={isFetching}
      onRefresh={dispatchFetchUsers}
    />
  );
}
