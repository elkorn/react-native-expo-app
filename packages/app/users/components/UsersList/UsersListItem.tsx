import React from "react";
import { StyleSheet, Text, View } from "react-native";
import { User } from "../../types";
import Avatar from "./Avatar";

export default function UsersListItem({ item }: UsersListItemProps) {
  return (
    <View style={styles.listItem}>
      <Avatar src={item.avatar} />
      <View>
        <Text style={styles.name}>{item.name}</Text>
        <Text>ID: {item.id}</Text>
      </View>
    </View>
  );
}

interface UsersListItemProps {
  item: User;
}

const styles = StyleSheet.create({
  listItem: {
    paddingVertical: 12,
    paddingHorizontal: 12,
    borderBottomWidth: 0.3,
    borderColor: "#ddd",
    flex: 1,
    flexDirection: "row",
    alignItems: "center",
  },

  name: {
    fontWeight: "bold",
    fontSize: 24,
  },
});
