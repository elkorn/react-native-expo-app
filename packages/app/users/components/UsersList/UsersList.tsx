import React from "react";
import { FlatList, RefreshControl } from "react-native";
import { User } from "../../types";
import UsersListItem from "./UsersListItem";
import { LoadingState } from "../../../shared/types";

export function UsersList({
  users = [],
  loadingState,
  onRefresh,
}: UsersListProps) {
  return (
    <FlatList
      refreshControl={
        <RefreshControl
          refreshing={loadingState === LoadingState.Loading}
          onRefresh={onRefresh}
        />
      }
      data={users}
      keyExtractor={(user) => user.id}
      renderItem={(props) => <UsersListItem {...props} />}
    />
  );
}

export interface UsersListProps {
  users?: User[];
  loadingState: LoadingState;
  onRefresh: () => void;
}
