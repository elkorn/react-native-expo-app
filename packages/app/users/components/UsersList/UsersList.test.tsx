import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import { UsersList, UsersListProps } from "./UsersList";
import { mockUsers } from "../../../shared/mocks/mockUsers";
import { LoadingState } from "../../../shared/types";

describe("components", () => {
  describe("UsersList", () => {
    describe("UsersList", () => {
      const baseProps: UsersListProps = {
        users: mockUsers,
        loadingState: LoadingState.Idle,
        onRefresh: () => {},
      };

      describe("matches snapshot", () => {
        it("loaded state", () => {
          const tree = renderer.create(<UsersList {...baseProps} />);
          expect(tree).toMatchSnapshot();
        });

        it("loading state", () => {
          const props = {
            ...baseProps,
            loadingState: LoadingState.Loading,
          };

          const tree = renderer.create(<UsersList {...props} />);
          expect(tree).toMatchSnapshot();
        });
      });
    });
  });
});
