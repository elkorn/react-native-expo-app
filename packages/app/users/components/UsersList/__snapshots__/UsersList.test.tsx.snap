// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components UsersList UsersList matches snapshot loaded state 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "avatar": " https://i.pravatar.cc/150?img=1",
        "id": "1",
        "name": "Morgan James",
      },
      Object {
        "avatar": "https://i.pravatar.cc/150?img=2",
        "id": "2",
        "name": "Esme Coleman",
      },
      Object {
        "avatar": "https://i.pravatar.cc/150?img=3",
        "id": "3",
        "name": "Gary Watson",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={false}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=1",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Morgan James
          </Text>
          <Text>
            ID: 
            1
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=2",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Esme Coleman
          </Text>
          <Text>
            ID: 
            2
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=3",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Gary Watson
          </Text>
          <Text>
            ID: 
            3
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;

exports[`components UsersList UsersList matches snapshot loading state 1`] = `
<RCTScrollView
  data={
    Array [
      Object {
        "avatar": " https://i.pravatar.cc/150?img=1",
        "id": "1",
        "name": "Morgan James",
      },
      Object {
        "avatar": "https://i.pravatar.cc/150?img=2",
        "id": "2",
        "name": "Esme Coleman",
      },
      Object {
        "avatar": "https://i.pravatar.cc/150?img=3",
        "id": "3",
        "name": "Gary Watson",
      },
    ]
  }
  disableVirtualization={false}
  getItem={[Function]}
  getItemCount={[Function]}
  horizontal={false}
  initialNumToRender={10}
  keyExtractor={[Function]}
  maxToRenderPerBatch={10}
  onContentSizeChange={[Function]}
  onEndReachedThreshold={2}
  onLayout={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  refreshControl={
    <RefreshControlMock
      onRefresh={[Function]}
      refreshing={true}
    />
  }
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={Array []}
  updateCellsBatchingPeriod={50}
  viewabilityConfigCallbackPairs={Array []}
  windowSize={21}
>
  <RCTRefreshControl />
  <View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=1",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Morgan James
          </Text>
          <Text>
            ID: 
            1
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=2",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Esme Coleman
          </Text>
          <Text>
            ID: 
            2
          </Text>
        </View>
      </View>
    </View>
    <View
      onLayout={[Function]}
      style={null}
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "borderBottomWidth": 0.3,
            "borderColor": "#ddd",
            "flex": 1,
            "flexDirection": "row",
            "paddingHorizontal": 12,
            "paddingVertical": 12,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "borderRadius": 100,
              "height": 100,
              "justifyContent": "center",
              "marginRight": 12,
              "position": "relative",
              "width": 100,
            }
          }
        >
          <Image
            onLoad={[Function]}
            onLoadStart={[Function]}
            source={
              Object {
                "uri": "https://i.pravatar.cc/150?img=3",
              }
            }
            style={
              Object {
                "backgroundColor": "#eee",
                "borderRadius": 100,
                "height": 100,
                "left": 0,
                "opacity": 0,
                "position": "absolute",
                "top": 0,
                "width": 100,
              }
            }
          />
          <ActivityIndicator
            animating={false}
            color="#999999"
            hidesWhenStopped={true}
            size="small"
          />
        </View>
        <View>
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
              }
            }
          >
            Gary Watson
          </Text>
          <Text>
            ID: 
            3
          </Text>
        </View>
      </View>
    </View>
  </View>
</RCTScrollView>
`;
